numba/core/errors.py

```diff
@@ -425,20 +426,52 @@ def deprecate_moved_module(old_module, new_module, stacklevel=3):
         msg = _moved_no_replacement.format(old_module)
     else:
         msg = _moved_msg1.format(old_module, new_module)
-    warnings.warn(msg, category=NumbaDeprecationWarning, stacklevel=stacklevel)
+    warnings.warn(
+        msg, category=NumbaDeprecationWarning, stacklevel=stacklevel + 1)
 
 
-def deprecate_moved_module_getattr(old_module, new_module):
-    """For replacing module level __getattr__ to warn users above modules moving
-    locations
-    """
-    def wrapper(attr):
-        mod = import_module(new_module)
-        ret_attr = getattr(mod, attr)
-        msg = _moved_msg2.format(attr, old_module, new_module)
-        warnings.warn(msg, category=NumbaDeprecationWarning, stacklevel=2)
-        return ret_attr
-    return wrapper
+class _MovedModule(ModuleType):
+    def __init__(self, old_module_locals, new_module):
+        old_module = old_module_locals['__name__']
+        super().__init__(old_module)
+
+        # copy across dunders so that package imports work too
+        for attr, value in old_module_locals.items():
+            if attr.startswith('__') and attr.endswith('__'):
+                setattr(self, attr, value)
+
+        self.__new_module = new_module
+        deprecate_moved_module(old_module, new_module, stacklevel=3)
+
+    def __getattr__(self, attr):
+        """ warn users above modules moving locations """
+
+        try:
+            # import from the moved module
+            if self.__new_module is not None:
+                mod = import_module(self.__new_module)
+                ret_attr = getattr(mod, attr)
+                msg = _moved_msg2.format(attr, self.__name__, self.__new_module)
+                warnings.warn(
+                    msg, category=NumbaDeprecationWarning, stacklevel=2)
+                return ret_attr
+            else:
+                # produce the usual error
+                return super().__getattribute__(attr)
+        except AttributeError:
+            # not a package, so no submodules to attempt to import.
+            # can't use hasattr here because that would recurse.
+            if '__path__' not in self.__dict__:
+                raise
+
+            # perhaps this is a submodule name that was previous importer, but
+            # is no longer
+            try:
+                return import_module("." + attr, package=self.__name__)
+            except ModuleNotFoundError:
+                raise AttributeError(
+                    "Moved module {!r} has no attribute or submodule {!r}"
+                    .format(self.__name__, attr))
```

Comment:
No test appears to hit this branch? I think something like this would exercise it?:
```
python -c "import numba.targets; numba.targets.does_not_exist"
```

Reply:
This branch is hit only sometimes, depending on which order the tests run in.

Starting a new interpreter would be the best way to ensure the test is hit, yes. Where would such a test go?

Reply:
Yes, and also perhaps capture the output and ensure that it contains the expected? This shim is intended to be short-lived in the code base, so don't worry about code coverage. I think it'd be fine to add the test to the `numba/tests/test_refactor_moves.py` given it's testing part of the same functionality. Thanks for doing this.

Reply:
Apologies, I didn't read this message properly, adding a test is straightforward